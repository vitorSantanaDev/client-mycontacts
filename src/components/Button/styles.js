import styled, { css } from 'styled-components'

export const StyledButton = styled.button`
	${({ theme, danger }) => css`
		border: none;
		height: 5.2rem;
		color: white;
		font-weight: ${theme.font.bold};
		border-radius: ${theme.border.radius};
		font-size: ${theme.font.sizes.medium};
		box-shadow: 0px 4px 10px rgba(0, 0, 0, 0.04);
		background-color: ${danger
			? theme.colors.danger.main
			: theme.colors.primary.main};

		padding: ${theme.spacings.xsmall};

		transition: background 0.2s ease-in;

		display: flex;
		align-items: center;
		justify-content: center;

		&:hover {
			background-color: ${danger
				? theme.colors.danger.light
				: theme.colors.primary.light};
		}

		&:active {
			background-color: ${danger
				? theme.colors.danger.dark
				: theme.colors.primary.dark};
		}

		&:disabled {
			background-color: #ccc !important;
			cursor: not-allowed !important;
		}
	`}
`
