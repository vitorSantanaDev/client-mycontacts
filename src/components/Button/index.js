import PropTypes from 'prop-types'

import { StyledButton } from './styles'
import Spinner from '../Spinner'

export default function Button({
	type,
	disabled,
	isLoading,
	children,
	onClick,
	danger
}) {
	return (
		<StyledButton
			danger={danger}
			type={type}
			onClick={onClick}
			disabled={disabled || isLoading}
		>
			{!isLoading && children}
			{isLoading && <Spinner size={16} />}
		</StyledButton>
	)
}

Button.propTypes = {
	danger: PropTypes.bool,
	type: PropTypes.string,
	disabled: PropTypes.bool,
	isLoading: PropTypes.bool,
	children: PropTypes.node.isRequired,
	onClick: PropTypes.func
}

Button.defaultProps = {
	type: 'button',
	disabled: false,
	isLoading: false,
	onClick: () => {},
	danger: false
}
