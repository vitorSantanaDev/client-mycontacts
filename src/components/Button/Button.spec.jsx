import React from 'react'

import Button from '.'
import renderWithTheme from '../../utils/render-with-theme'
import { screen } from '@testing-library/react'
import defaultTheme from '../../assets/styles/themes/default'
import userEvent from '@testing-library/user-event'

jest.mock('../Spinner', () => {
	return {
		__esModule: true,
		default: function Mock({ size }) {
			return (
				<div
					data-testid="Mock Spinner Component"
					style={{ fontSize: size }}
				></div>
			)
		}
	}
})

describe('<Button />', () => {
	it('should render button with default props', () => {
		renderWithTheme(<Button>Button</Button>)

		const button = screen.getByRole('button', { name: /button/i })

		expect(button).toBeInTheDocument()
		expect(button.getAttribute('type')).toStrictEqual('button')
		expect(button).toHaveStyle({
			backgroundColor: defaultTheme.colors.primary.main
		})
	})

	it('should render button with background danger, when prop danger is true', () => {
		renderWithTheme(<Button danger>Button</Button>)

		const button = screen.getByRole('button', { name: /button/i })

		expect(button).toHaveStyle({
			backgroundColor: defaultTheme.colors.danger.main
		})
	})

	it('should render the button with a spinner when prop loading is true', () => {
		renderWithTheme(<Button isLoading>Button</Button>)

		const spinner = screen.getByTestId('Mock Spinner Component')

		const button = spinner.parentElement

		expect(spinner).toBeInTheDocument()

		expect(button).toHaveStyle({
			backgroundColor: '#ccc',
			cursor: 'not-allowed'
		})
	})

	it('should call the handleClick function when the button is clicked', () => {
		const handleClick = jest.fn()
		renderWithTheme(<Button onClick={handleClick}>Button</Button>)

		const button = screen.getByRole('button', { name: /button/i })
		userEvent.click(button)

		expect(handleClick).toBeCalledTimes(1)
	})
})
