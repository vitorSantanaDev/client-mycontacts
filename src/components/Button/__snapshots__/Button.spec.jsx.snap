// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
.c0 {
  border: none;
  height: 5.2rem;
  color: white;
  font-weight: 700;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  box-shadow: 0px 4px 10px rgba(0,0,0,0.04);
  background-color: #5061FC;
  padding: 1.6rem;
  -webkit-transition: background 0.2s ease-in;
  transition: background 0.2s ease-in;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:hover {
  background-color: #6674f4;
}

.c0:active {
  background-color: #3346f0;
}

.c0:disabled {
  background-color: #ccc !important;
  cursor: not-allowed !important;
}

<button
  class="c0"
  type="button"
>
  Button
</button>
`;
